module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,58 @@
+#include "compression.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+
+CompressedBitmap compressGrayscale(const Bitmap& bitmap) {
+    CompressedBitmap compressed;
+    compressed.reserve(width * height);
+
+    int row{};
+
+    do {
+        for (auto beg = bitmap[row].begin(),
+                  end = bitmap[row].end();
+             beg != end;) {
+            const auto diff = std::find_if(beg, end,
+                                           [&beg](const auto& el) {
+                                               return el != *beg;
+                                           });
+            compressed.emplace_back(*beg, std::distance(beg, diff));
+            beg = diff;
+        }
+    } while (++row != height);
+
+    compressed.shrink_to_fit();
+    return compressed;
+}
+
+Bitmap decompressGrayscale(const CompressedBitmap& compressed) {
+    Bitmap bitmap;
+    auto row = bitmap.begin();
+    auto col = row->begin();
+
+    for (const auto& pair : compressed) {
+        if (col >= row->end()) {
+            std::advance(row, 1);
+            col = row->begin();
+        }
+        std::fill_n(col, pair.second, pair.first);
```

Comment:
I think it is possible to do that and advance will not be necessary. If you have time you can try
```suggestion
        col = std::fill_n(col, pair.second, pair.first);
```

Reply:
Indeed, I didn't check what was the return value of 'std::fill_n'. Your solution was very useful Kasperice - thank you! I also eliminated the previous occurrence of 'std::advanced' and replaced with simple 'row++'. The tests passed.

Reply:
Welcome :) So I remembered correctly ;)